File: scripts/fill-series-gaps.py
#!/usr/bin/env python3
"""Fill missing books in series that have gaps in position numbering."""

import html
import os
import re
import sqlite3
import time
import unicodedata
import uuid
from collections import Counter

import requests
from dotenv import load_dotenv

load_dotenv('.env.local')

BRAVE_API_KEY = os.getenv('BRAVE_SEARCH_API_KEY')
GOOGLE_BOOKS_API_KEY = os.getenv('GOOGLE_BOOKS_API_KEY')

DB_PATH = 'data/tbra.db'
GOOGLE_BOOKS_CAP = 900
google_books_calls = 0

stats = {'filled': 0, 'skipped': 0}


def normalize_title(title: str) -> str:
    t = title.lower().strip()
    t = unicodedata.normalize('NFKD', t)
    t = re.sub(r'[^\w\s]', '', t)
    t = re.sub(r'\s+', ' ', t)
    return t


def strip_html(text: str) -> str:
    text = re.sub(r'<[^>]+>', '', text)
    text = html.unescape(text)
    return text.strip()


def extract_clean_title(raw: str, series_name: str) -> str | None:
    """Extract just the book title from a raw match, stripping everything after it."""
    t = strip_html(raw).strip()

    # Cut at first occurrence of common terminators
    # Year in parens: (2019), (1984)
    t = re.split(r'\s*\(\d{4}\)', t)[0].strip()
    # Next numbered entry: "5." or "#5" or "Book 5"
    t = re.split(r'\s+\d+\.\s', t)[0].strip()
    t = re.split(r'\s+#\d', t)[0].strip()
    t = re.split(r'\s+Book\s+\d', t, flags=re.IGNORECASE)[0].strip()
    # Pipe, middot, or similar
    t = re.split(r'\s*[·|]\s*', t)[0].strip()
    # "by Author" pattern
    t = re.split(r'\s+by\s+[A-Z]', t)[0].strip()
    # Comma followed by number (next entry in comma-sep list)
    t = re.split(r',\s*\d', t)[0].strip()
    # Open paren (subtitle or metadata)
    # Only split if there's content before the paren
    parts = re.split(r'\s*\(', t)
    if parts[0].strip():
        t = parts[0].strip()
    # "..." or ellipsis
    t = re.split(r'\s*\.{2,}', t)[0].strip()
    t = re.split(r'\s*…', t)[0].strip()
    # " - " followed by subtitle or description
    # Be careful: some titles have " - " legitimately
    # Only strip if what follows looks like metadata
    m = re.match(r'^(.+?)\s+[-–—]\s+(?:Book|Vol|Part|Series|The\s+\w+\s+Saga)', t)
    if m:
        t = m.group(1).strip()

    # Final cleanup
    t = t.strip(' -–—|:;,*"\'()[]')
    # Remove trailing " * " artifacts
    t = re.sub(r'\s*\*\s*$', '', t).strip()

    if not t or len(t) < 3 or len(t) > 80:
        return None
    if not t[0].isalpha():
        return None
    alpha = sum(1 for c in t if c.isalpha())
    if alpha < 3:
        return None
    if normalize_title(t) == normalize_title(series_name):
        return None

    # Reject junk
    junk = [
        r'amazon', r'\.com', r'reviews?\b', r'ISBN\b', r'PDF\b',
        r'collects?\b', r'published\b', r'editors?\b', r'learn\s+more',
        r'upvotes', r'comments\b', r'r/', r'announcement',
        r'sexy', r'charming', r'eagerly', r'anticipated', r'respected',
        r'originally', r'description\b', r'fortress', r'minneapolis',
        r'\d{10}', r'on\s+sale', r'buy\s+now', r'free\s+shipping',
        r'paperback\b', r'hardcover\b', r'kindle\b', r'audible\b',
        r'audiobook\b', r'price\b', r'\$\d', r'sassy\s+review',
        r'founding\s+author', r'start\s+reading', r'google\s+books',
        r'salvatore', r'riftwar', r'follow\b',
    ]
    for pat in junk:
        if re.search(pat, t, re.IGNORECASE):
            return None

    if ';' in t:
        return None

    # Reject likely author names (2 capitalized words, all alpha)
    words = t.split()
    if len(words) == 2 and all(w[0].isupper() and w.isalpha() for w in words):
        # Likely a person name unless contains book-ish words
        book_words = {'blood', 'fire', 'war', 'dark', 'light', 'night', 'storm',
                      'shadow', 'throne', 'king', 'queen', 'dragon', 'magic',
                      'death', 'dead', 'bone', 'iron', 'steel', 'glass', 'stone',
                      'sword', 'crown', 'tower', 'star', 'moon', 'sun', 'sea',
                      'wind', 'frost', 'ice', 'gold', 'silver', 'red', 'black',
                      'white', 'blue', 'green', 'wild', 'lost', 'last', 'first',
                      'heat', 'cold', 'ivory', 'secret', 'evil', 'holy', 'sacred'}
        if not any(w.lower() in book_words for w in words):
            return None

    return t


def find_series_with_gaps(conn):
    cur = conn.execute("""
        SELECT s.id, s.name,
               GROUP_CONCAT(CAST(bs.position_in_series AS INTEGER), ',') as positions
        FROM series s
        JOIN book_series bs ON bs.series_id = s.id
        WHERE bs.position_in_series IS NOT NULL
          AND CAST(bs.position_in_series AS INTEGER) = bs.position_in_series
          AND bs.position_in_series > 0
        GROUP BY s.id
        HAVING MAX(bs.position_in_series) <= 20
    """)

    gaps = []
    for row in cur.fetchall():
        series_id, series_name, pos_str = row
        positions = sorted(set(int(p) for p in pos_str.split(',')))
        if not positions:
            continue
        min_pos = positions[0]
        max_pos = positions[-1]
        expected = set(range(min_pos, max_pos + 1))
        missing = sorted(expected - set(positions))
        if 1 <= len(missing) <= 5:
            gaps.append((series_id, series_name, missing, positions))
    return gaps


def get_series_author(conn, series_id):
    cur = conn.execute("""
        SELECT ba.author_id, COUNT(*) as cnt
        FROM book_series bs
        JOIN book_authors ba ON ba.book_id = bs.book_id
        WHERE bs.series_id = ?
        GROUP BY ba.author_id
        ORDER BY cnt DESC
        LIMIT 1
    """, (series_id,))
    row = cur.fetchone()
    return row[0] if row else None


def title_exists(conn, title):
    norm = normalize_title(title)
    cur = conn.execute("SELECT title FROM books")
    for (existing,) in cur.fetchall():
        if normalize_title(existing) == norm:
            return True
    return False


def ol_key_exists(conn, ol_key):
    if not ol_key:
        return False
    cur = conn.execute("SELECT 1 FROM books WHERE open_library_key = ?", (ol_key,))
    return cur.fetchone() is not None


def brave_search(query):
    if not BRAVE_API_KEY:
        return None
    url = 'https://api.search.brave.com/res/v1/web/search'
    headers = {'X-Subscription-Token': BRAVE_API_KEY, 'Accept': 'application/json'}
    params = {'q': query, 'count': 10}
    try:
        r = requests.get(url, headers=headers, params=params, timeout=15)
        time.sleep(1)
        if r.status_code == 200:
            return r.json()
    except Exception as e:
        print(f"  Brave search error: {e}")
    return None


def extract_titles_from_brave(results, series_name, position):
    if not results or 'web' not in results:
        return []

    candidates = []
    pos = str(position)

    for item in results.get('web', {}).get('results', []):
        raw_title = strip_html(item.get('title', ''))
        raw_desc = strip_html(item.get('description', ''))

        for text in [raw_title, raw_desc]:
            # "Book N: Title" / "#N: Title" / "Book N - Title"
            for m in re.finditer(
                rf'(?:book|#)\s*{pos}\s*[:\-–—]\s*(.+)',
                text, re.IGNORECASE
            ):
                title = extract_clean_title(m.group(1), series_name)
                if title:
                    candidates.append(title)

            # "N. Title" in a numbered list
            for m in re.finditer(
                rf'(?:^|\s){pos}\.\s+([A-Z].+)', text
            ):
                title = extract_clean_title(m.group(1), series_name)
                if title:
                    candidates.append(title)

    return candidates


def search_open_library(title, author_name=None):
    params = {'title': title, 'limit': 5}
    if author_name:
        params['author'] = author_name
    try:
        r = requests.get('https://openlibrary.org/search.json', params=params, timeout=15)
        time.sleep(0.5)
        if r.status_code != 200:
            return {}
        data = r.json()
        docs = data.get('docs', [])
        if not docs:
            return {}

        norm_title = normalize_title(title)
        best = None
        for doc in docs:
            if normalize_title(doc.get('title', '')) == norm_title:
                best = doc
                break
        if not best:
            doc_norm = normalize_title(docs[0].get('title', ''))
            if norm_title in doc_norm or doc_norm in norm_title:
                best = docs[0]
            else:
                return {}

        result = {}
        if best.get('key'):
            result['open_library_key'] = best['key']
        if best.get('cover_i'):
            result['cover_image_url'] = f"https://covers.openlibrary.org/b/id/{best['cover_i']}-L.jpg"
        if best.get('number_of_pages_median'):
            result['pages'] = best['number_of_pages_median']
        if best.get('first_publish_year'):
            result['publication_year'] = best['first_publish_year']
        if best.get('isbn') and len(best['isbn']) > 0:
            for isbn in best['isbn']:
                if len(isbn) == 13:
                    result['isbn_13'] = isbn
                    break
            if 'isbn_13' not in result:
                for isbn in best['isbn']:
                    if len(isbn) == 10:
                        result['isbn_10'] = isbn
                        break
        if best.get('first_sentence'):
            sentences = best['first_sentence']
            if isinstance(sentences, list) and sentences:
                result['description'] = sentences[0]
        return result
    except Exception as e:
        print(f"  Open Library error: {e}")
        return {}


def search_google_books_cover(title, author_name=None):
    global google_books_calls
    if not GOOGLE_BOOKS_API_KEY or google_books_calls >= GOOGLE_BOOKS_CAP:
        return None
    q = f'intitle:{title}'
    if author_name:
        q += f' inauthor:{author_name}'
    params = {'q': q, 'key': GOOGLE_BOOKS_API_KEY, 'maxResults': 1}
    try:
        r = requests.get('https://www.googleapis.com/books/v1/volumes', params=params, timeout=15)
        google_books_calls += 1
        time.sleep(0.5)
        if r.status_code != 200:
            return None
        data = r.json()
        items = data.get('items', [])
        if items:
            info = items[0].get('volumeInfo', {})
            imgs = info.get('imageLinks', {})
            return imgs.get('thumbnail') or imgs.get('smallThumbnail')
    except Exception as e:
        print(f"  Google Books error: {e}")
    return None


def get_author_name(conn, author_id):
    cur = conn.execute("SELECT name FROM authors WHERE id = ?", (author_id,))
    row = cur.fetchone()
    return row[0] if row else None


def create_book(conn, title, metadata, series_id, position, author_id):
    book_id = str(uuid.uuid4())
    ol_key = metadata.get('open_library_key')
    if ol_key_exists(conn, ol_key):
        ol_key = None

    conn.execute("""
        INSERT INTO books (id, title, description, publication_year, isbn_10, isbn_13,
                           pages, cover_image_url, open_library_key, language, is_fiction,
                           created_at, updated_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 'English', 1, datetime('now'), datetime('now'))
    """, (
        book_id, title, metadata.get('description'), metadata.get('publication_year'),
        metadata.get('isbn_10'), metadata.get('isbn_13'), metadata.get('pages'),
        metadata.get('cover_image_url'), ol_key,
    ))

    if author_id:
        conn.execute("INSERT INTO book_authors (book_id, author_id, role) VALUES (?, ?, 'author')",
                     (book_id, author_id))

    conn.execute("INSERT INTO book_series (book_id, series_id, position_in_series) VALUES (?, ?, ?)",
                 (book_id, series_id, position))

    conn.commit()
    return book_id


def pick_best_title(candidates):
    if not candidates:
        return None
    counts = Counter(normalize_title(c) for c in candidates)
    best_norm, _ = counts.most_common(1)[0]
    for c in candidates:
        if normalize_title(c) == best_norm:
            return c
    return None


def main():
    os.chdir(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    conn = sqlite3.connect(DB_PATH)
    conn.execute("PRAGMA journal_mode=WAL")

    print("Finding series with gaps...")
    gaps = find_series_with_gaps(conn)
    print(f"Found {len(gaps)} series with gaps\n")

    for series_id, series_name, missing_positions, existing_positions in gaps:
        author_id = get_series_author(conn, series_id)
        author_name = get_author_name(conn, author_id) if author_id else None

        for position in missing_positions:
            all_candidates = []

            q1 = f'"{series_name}" series reading order'
            if author_name:
                q1 += f' {author_name}'
            r1 = brave_search(q1)
            all_candidates.extend(extract_titles_from_brave(r1, series_name, position))

            if len(all_candidates) < 2:
                q2 = f'"{series_name}" book {position}'
                if author_name:
                    q2 += f' {author_name}'
                r2 = brave_search(q2)
                all_candidates.extend(extract_titles_from_brave(r2, series_name, position))

            title = pick_best_title(all_candidates)

            if not title:
                print(f"[series-gap] SKIP {series_name} #{position}: could not identify title")
                stats['skipped'] += 1
                continue

            if title_exists(conn, title):
                print(f"[series-gap] SKIP {series_name} #{position}: '{title}' already exists")
                stats['skipped'] += 1
                continue

            ol_meta = search_open_library(title, author_name)

            if not ol_meta.get('cover_image_url'):
                gb_cover = search_google_books_cover(title, author_name)
                if gb_cover:
                    ol_meta['cover_image_url'] = gb_cover

            try:
                create_book(conn, title, ol_meta, series_id, position, author_id)
                print(f"[series-gap] Filling {series_name} #{position}: {title}")
                stats['filled'] += 1
            except Exception as e:
                print(f"[series-gap] ERROR {series_name} #{position}: {title} - {e}")
                conn.rollback()
                stats['skipped'] += 1

    conn.close()

    print(f"\n{'='*50}")
    print(f"Summary:")
    print(f"  Total filled:          {stats['filled']}")
    print(f"  Total skipped:         {stats['skipped']}")
    print(f"  Google Books calls:    {google_books_calls}")


if __name__ == '__main__':
    main()

File: scripts/fill-series-gaps.ts
/**
 * fill-series-gaps.ts — Overnight script to find and add missing books in series
 *
 * Uses Brave Search to identify missing books, then OpenLibrary to import them.
 * Enrichment runs via the standard pipeline (no Brave for enrichment itself).
 *
 * Usage: npx tsx scripts/fill-series-gaps.ts
 */

import { config } from "dotenv";
config({ path: ".env.local" });

import Database from "better-sqlite3";
import path from "path";
import crypto from "crypto";

const DB_PATH = path.join(process.cwd(), "data", "tbra.db");
const db = new Database(DB_PATH);
db.pragma("journal_mode = WAL");
db.pragma("foreign_keys = ON");

const BRAVE_API_KEY = process.env.BRAVE_SEARCH_API_KEY!;
const BRAVE_DELAY_MS = 1200; // stay well under rate limit
const OL_DELAY_MS = 500;

interface GapEntry {
  seriesId: string;
  seriesName: string;
  position: number;
  author: string;
  userCount: number;
}

interface OLWork {
  key: string;
  title: string;
  cover_i?: number;
  first_publish_year?: number;
  number_of_pages_median?: number;
  author_name?: string[];
  isbn?: string[];
}

// ─── Load gaps from database ───
function loadGaps(): GapEntry[] {
  const rows = db.prepare(`
    SELECT s.id as series_id, s.name as series_name,
           GROUP_CONCAT(DISTINCT CAST(bs.position_in_series AS INTEGER)) as positions,
           MAX(CAST(bs.position_in_series AS INTEGER)) as max_pos,
           AVG(b.pages) as avg_pages,
           COUNT(DISTINCT bs.book_id) as book_count,
           COUNT(DISTINCT ubs.user_id) as user_count
    FROM series s
    JOIN book_series bs ON bs.series_id = s.id
    JOIN books b ON b.id = bs.book_id
    LEFT JOIN user_book_state ubs ON ubs.book_id = bs.book_id
    WHERE bs.position_in_series IS NOT NULL
      AND CAST(bs.position_in_series AS INTEGER) = bs.position_in_series
      AND bs.position_in_series > 0
      AND b.is_box_set = 0
    GROUP BY s.id
    HAVING book_count >= 2 AND max_pos <= 20
    ORDER BY user_count DESC, book_count DESC
  `).all() as any[];

  const gaps: GapEntry[] = [];

  for (const row of rows) {
    const positions = new Set(row.positions.split(",").map(Number));
    const maxPos = row.max_pos;
    if (row.avg_pages && row.avg_pages < 100) continue; // skip comics

    // Get author
    const authorRow = db.prepare(`
      SELECT DISTINCT a.name FROM authors a
      JOIN book_authors ba ON ba.author_id = a.id
      JOIN book_series bs ON bs.book_id = ba.book_id
      WHERE bs.series_id = ? LIMIT 1
    `).get(row.series_id) as any;
    const author = authorRow?.name || "unknown";

    for (let pos = 1; pos <= maxPos; pos++) {
      if (!positions.has(pos)) {
        gaps.push({
          seriesId: row.series_id,
          seriesName: row.series_name,
          position: pos,
          author,
          userCount: row.user_count,
        });
      }
    }

    // Max 5 gaps per series
    const seriesGaps = gaps.filter(g => g.seriesId === row.series_id);
    if (seriesGaps.length > 5) {
      // Remove excess
      const excess = seriesGaps.slice(5);
      for (const e of excess) {
        const idx = gaps.indexOf(e);
        if (idx >= 0) gaps.splice(idx, 1);
      }
    }
  }

  return gaps;
}

// ─── Brave Search ───
async function braveSearch(seriesName: string, position: number, author: string): Promise<string | null> {
  // Multiple search strategies, from most to least specific
  const queries = [
    `"${seriesName}" "book ${position}" ${author} title`,
    `${seriesName} #${position} ${author} novel goodreads`,
    `${seriesName} series book ${position} ${author}`,
  ];

  for (const query of queries) {
    try {
      const url = `https://api.search.brave.com/res/v1/web/search?q=${encodeURIComponent(query)}&count=8`;
      const res = await fetch(url, {
        headers: { "X-Subscription-Token": BRAVE_API_KEY },
      });
      if (!res.ok) {
        console.error(`  Brave error: ${res.status}`);
        await sleep(BRAVE_DELAY_MS);
        continue;
      }
      const data = await res.json();
      const results = data.web?.results || [];

      // Look through results for a specific book title (not the series name itself)
      for (const r of results) {
        const title = r.title || "";
        const desc = r.description || "";
        const combined = `${title} ${desc}`.toLowerCase();

        // Skip results that are just about the series, not a specific book
        if (title.toLowerCase().trim() === seriesName.toLowerCase().trim()) continue;
        if (title.toLowerCase().includes("series") && !title.toLowerCase().includes("book")) continue;

        // Look for Goodreads or Amazon links which tend to have clean book titles
        if (r.url?.includes("goodreads.com/book/") || r.url?.includes("amazon.com")) {
          // Check if it mentions the right position
          if (combined.includes(`#${position}`) || combined.includes(`book ${position}`) ||
              combined.includes(`vol. ${position}`) || combined.includes(`volume ${position}`)) {
            // Extract title: "Book Title (Series, #N)" → "Book Title"
            let cleanTitle = title
              .replace(/\s*\(.*\)\s*$/, "")
              .replace(/\s*\|.*$/, "")
              .replace(/\s*:\s*Amazon.*$/i, "")
              .replace(/\s*by\s+.*$/i, "")
              .replace(/\s*-\s*Kindle.*$/i, "")
              .trim();
            if (cleanTitle.length > 3 && cleanTitle.length < 200) {
              return cleanTitle;
            }
          }
        }
      }

      await sleep(BRAVE_DELAY_MS);
    } catch (err) {
      console.error(`  Brave search error:`, err);
    }
  }

  return null;
}

// ─── OpenLibrary Search ───
async function searchOpenLibrary(title: string, author: string): Promise<OLWork | null> {
  try {
    const query = `${title} ${author}`.trim();
    const url = `https://openlibrary.org/search.json?q=${encodeURIComponent(query)}&limit=5&fields=key,title,cover_i,first_publish_year,number_of_pages_median,author_name,isbn`;
    const res = await fetch(url);
    if (!res.ok) return null;
    const data = await res.json();

    if (!data.docs || data.docs.length === 0) return null;

    // Find best match — prefer one with cover and matching author
    const authorLower = author.toLowerCase();
    const titleLower = title.toLowerCase();

    for (const doc of data.docs) {
      const docTitle = (doc.title || "").toLowerCase();
      const docAuthors = (doc.author_name || []).map((a: string) => a.toLowerCase());

      // Title should be similar
      if (!docTitle.includes(titleLower.slice(0, 10)) && !titleLower.includes(docTitle.slice(0, 10))) continue;

      // Author should match
      if (docAuthors.some((a: string) => a.includes(authorLower.split(" ").pop()!))) {
        return doc;
      }
    }

    // Fallback: first result if title matches reasonably
    const first = data.docs[0];
    if (first && (first.title || "").toLowerCase().includes(titleLower.slice(0, 8))) {
      return first;
    }

    return null;
  } catch (err) {
    console.error(`  OL search error:`, err);
    return null;
  }
}

// ─── Check if book already exists ───
function bookExists(olKey: string | null, title: string, author: string): boolean {
  if (olKey) {
    const existing = db.prepare("SELECT id FROM books WHERE open_library_key = ?").get(olKey);
    if (existing) return true;
  }

  // Fuzzy title+author check
  const normalized = title.toLowerCase().replace(/[^a-z0-9]/g, "");
  const existing = db.prepare(`
    SELECT b.id FROM books b
    JOIN book_authors ba ON ba.book_id = b.id
    JOIN authors a ON a.id = ba.author_id
    WHERE LOWER(REPLACE(REPLACE(REPLACE(b.title, ' ', ''), '-', ''), '''', '')) LIKE ?
    AND LOWER(a.name) LIKE ?
    LIMIT 1
  `).get(`%${normalized.slice(0, 20)}%`, `%${author.split(" ").pop()?.toLowerCase()}%`);

  return !!existing;
}

// ─── Create slug ───
function createSlug(title: string, author: string): string {
  const base = `${title}-${author}`
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, "")
    .replace(/\s+/g, "-")
    .replace(/-+/g, "-")
    .slice(0, 80);
  return base;
}

// ─── Insert book ───
function insertBook(
  title: string,
  author: string,
  olWork: OLWork | null,
  seriesId: string,
  position: number
): string | null {
  const bookId = crypto.randomUUID();
  const authorId = crypto.randomUUID();
  const slug = createSlug(title, author);

  const olKey = olWork?.key?.replace("/works/", "") || null;
  const coverUrl = olWork?.cover_i
    ? `https://covers.openlibrary.org/b/id/${olWork.cover_i}-L.jpg`
    : null;
  const year = olWork?.first_publish_year || null;
  const pages = olWork?.number_of_pages_median || null;
  let isbn13 = olWork?.isbn?.find((i: string) => i.length === 13) || null;

  // Check if ISBN already exists (different edition of same book)
  if (isbn13) {
    const existingIsbn = db.prepare("SELECT id FROM books WHERE isbn_13 = ?").get(isbn13);
    if (existingIsbn) isbn13 = null; // skip ISBN to avoid conflict
  }

  // Check if already exists
  if (bookExists(olKey, title, author)) {
    // Book exists — just link to series if not already linked
    let existingId: string | null = null;
    if (olKey) {
      const row = db.prepare("SELECT id FROM books WHERE open_library_key = ?").get(olKey) as any;
      if (row) existingId = row.id;
    }
    if (!existingId) {
      const normalized = title.toLowerCase().replace(/[^a-z0-9]/g, "");
      const row = db.prepare(`
        SELECT b.id FROM books b
        JOIN book_authors ba ON ba.book_id = b.id
        JOIN authors a ON a.id = ba.author_id
        WHERE LOWER(REPLACE(REPLACE(REPLACE(b.title, ' ', ''), '-', ''), '''', '')) LIKE ?
        AND LOWER(a.name) LIKE ?
        LIMIT 1
      `).get(`%${normalized.slice(0, 20)}%`, `%${author.split(" ").pop()?.toLowerCase()}%`) as any;
      if (row) existingId = row.id;
    }

    if (existingId) {
      // Check if series link exists
      const link = db.prepare(
        "SELECT 1 FROM book_series WHERE book_id = ? AND series_id = ?"
      ).get(existingId, seriesId);
      if (!link) {
        db.prepare(
          "INSERT INTO book_series (book_id, series_id, position_in_series) VALUES (?, ?, ?)"
        ).run(existingId, seriesId, position);
        console.log(`  ↳ Linked existing book to series at position ${position}`);
      } else {
        console.log(`  ↳ Already linked to series`);
      }
      return existingId;
    }
    return null;
  }

  // Insert new book
  db.prepare(`
    INSERT INTO books (id, title, slug, open_library_key, cover_image_url, publication_year,
                       pages, isbn_13, visibility, is_box_set)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, 'public', 0)
  `).run(bookId, title, slug, olKey, coverUrl, year, pages, isbn13);

  // Find or create author
  let existingAuthor = db.prepare(
    "SELECT id FROM authors WHERE LOWER(name) = LOWER(?)"
  ).get(author) as any;

  let finalAuthorId = existingAuthor?.id;
  if (!finalAuthorId) {
    const authorSlug = author.toLowerCase().replace(/[^a-z0-9\s-]/g, "").replace(/\s+/g, "-");
    db.prepare("INSERT INTO authors (id, name, slug) VALUES (?, ?, ?)").run(authorId, author, authorSlug);
    finalAuthorId = authorId;
  }

  // Link book to author
  db.prepare("INSERT INTO book_authors (book_id, author_id) VALUES (?, ?)").run(bookId, finalAuthorId);

  // Link book to series
  db.prepare(
    "INSERT INTO book_series (book_id, series_id, position_in_series) VALUES (?, ?, ?)"
  ).run(bookId, seriesId, position);

  // Add enrichment log entry so nightly enrichment picks it up
  db.prepare(`
    INSERT INTO enrichment_log (id, book_id, status, source, started_at)
    VALUES (?, ?, 'pending', 'series-gap-fill', datetime('now'))
  `).run(crypto.randomUUID(), bookId);

  return bookId;
}

// ─── Main ───
async function main() {
  const gaps = loadGaps();
  console.log(`\n📚 Series Gap Filler — ${gaps.length} missing books across ${new Set(gaps.map(g => g.seriesId)).size} series\n`);

  let found = 0;
  let linked = 0;
  let notFound = 0;
  let errors = 0;

  // Group by series for nicer output
  let currentSeries = "";

  for (let i = 0; i < gaps.length; i++) {
    const gap = gaps[i];

    if (gap.seriesName !== currentSeries) {
      currentSeries = gap.seriesName;
      const userTag = gap.userCount > 0 ? ` [${gap.userCount} users]` : "";
      console.log(`\n─── ${gap.seriesName} by ${gap.author}${userTag} ───`);
    }

    console.log(`  Searching for book #${gap.position}...`);

    // Search Brave for the book title
    const braveTitle = await braveSearch(gap.seriesName, gap.position, gap.author);

    if (!braveTitle) {
      console.log(`  ✗ No results from Brave`);
      notFound++;
      continue;
    }

    console.log(`  Brave found: "${braveTitle}"`);

    // Search OpenLibrary for the book
    const olWork = await searchOpenLibrary(braveTitle, gap.author);
    await sleep(OL_DELAY_MS);

    if (!olWork) {
      console.log(`  ✗ Not found on OpenLibrary`);
      notFound++;
      continue;
    }

    console.log(`  OL match: "${olWork.title}" (${olWork.first_publish_year || "?"}) — ${olWork.number_of_pages_median || "?"} pages`);

    try {
      const bookId = insertBook(olWork.title, gap.author, olWork, gap.seriesId, gap.position);
      if (bookId) {
        found++;
        console.log(`  ✓ Added as position #${gap.position}`);
      } else {
        linked++;
      }
    } catch (err: any) {
      console.error(`  ✗ Insert error: ${err.message}`);
      errors++;
    }
  }

  console.log(`\n${"═".repeat(50)}`);
  console.log(`Done! Found: ${found}, Linked existing: ${linked}, Not found: ${notFound}, Errors: ${errors}`);
  console.log(`Total: ${found + linked} books added/linked to series`);
  console.log(`${found} new books queued for nightly enrichment`);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

main().catch(console.error);
